v3.0 edits (#3743)
viamrobotics/docs: docs/data-ai/ai/run-inference.md

weight: 50
layout: "docs"
type: "docs"
modulescript: true
description: "Run inference on a model with a vision service or an SDK."
---

After deploying an ml model, you need to configure an additional service to use the inferences the deployed model makes.

> You can run inference on an ML model with a vision service in the app or use an SDK.

Reviewer: ```suggestion
You can run inference on an ML model with a vision service or use an SDK to further process inferences.
```

## Use a vision service

Vision services work to provide computer vision.
They use an ML model and apply it to the stream of images from your camera.

{{<resources_svc api="rdk:service:vision" type="vision">}}

{{< readfile "/static/include/create-your-own-mr.md" >}}

Note that many of these services have built in ML models, and thus do not need to be run alongside an ML model service.

One vision service you can use to run inference on a camera stream if you have an ML model service configured is the `mlmodel` service.

> ### Configure an mlmodel vision service
> 
> Add the `vision / ML model` service to your machine.
> Then, from the **Select model** dropdown, select the name of the ML model service you configured when [deploying](/data-ai/ai/deploy/) your model (for example, `mlmodel-1`).
> 
> **Save** your changes.
> 
> ### Test your changes

Reviewer: ```suggestion
### Check vision service 
```

Author: sorry-- why not have this info on here? is it too self-explanatory?

> You can test your vision service by clicking on the **Test** area of its configuration panel or from the [**CONTROL** tab](/fleet/control/).

Reviewer: ```suggestion
You can test a deployed vision service by clicking on the **Test** area of its configuration panel or from the [**CONTROL** tab](/fleet/control/).
```

Try pointing the camera at a scene similar to your training data.

{{< imgproc src="/tutorials/data-management/blue-star.png" alt="Detected blue star" resize="x200" >}}
{{< imgproc src="/tutorials/filtered-camera-module/viam-figure-preview.png" alt="Detection of a viam figure with a confidence score of 0.97" resize="x200" >}}

{{% expand "Want to limit the number of shown classifications or detections? Click here." %}}

If you are seeing a lot of classifications or detections, you can set a minimum confidence threshold.

Start by setting the value to 0.8.
This reduces your output by filtering out anything below a threshold of 80% confidence.
You can adjust this attribute as necessary.

Click the **Save** button in the top right corner of the page to save your configuration, then close and reopen the **Test** panel of the vision service configuration panel.
Now if you reopen the panel, you will only see classifications or detections with a confidence value higher than the `default_minimum_confidence` attribute.

{{< /expand>}}

For more detailed information, including optional attribute configuration, see the [`mlmodel` docs](/services/vision/mlmodel/).

> ## Use an SDK

Reviewer: Good start, I think we could possibly add a code sample here before pointing to APIs?
